File: clients/go/rpc_test.go
package oresapidocs

import (
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"testing"
)

type fixtureDocument struct {
	SchemaVersion int    `json:"schemaVersion"`
	Profile       string `json:"profile"`
	Valid         []struct {
		Name         string `json:"name"`
		Kind         string `json:"kind"`
		Encoded      string `json:"encoded"`
		TCPPrefixHex string `json:"tcp_prefix_hex"`
	} `json:"valid"`
	Invalid []struct {
		Name    string `json:"name"`
		Kind    string `json:"kind"`
		Encoded string `json:"encoded"`
	} `json:"invalid"`
}

func TestSharedFixturesRoundTrip(t *testing.T) {
	data, err := os.ReadFile(filepath.FromSlash("../../examples/rpc-v1/conformance.json"))
	if err != nil {
		t.Fatal(err)
	}
	var document fixtureDocument
	if err := json.Unmarshal(data, &document); err != nil {
		t.Fatal(err)
	}
	if document.SchemaVersion != 1 || document.Profile != "ores-rpc-v1-call-receipt" {
		t.Fatal("fixture profile drift")
	}
	for _, fixture := range document.Valid {
		t.Run(fixture.Name, func(t *testing.T) {
			var encoded []byte
			if fixture.Kind == "call" {
				value, err := DecodeCall([]byte(fixture.Encoded))
				if err != nil {
					t.Fatal(err)
				}
				encoded, err = value.Encode()
				if err != nil {
					t.Fatal(err)
				}
			} else {
				value, err := DecodeReceipt([]byte(fixture.Encoded))
				if err != nil {
					t.Fatal(err)
				}
				encoded, err = value.Encode()
				if err != nil {
					t.Fatal(err)
				}
			}
			if string(encoded) != fixture.Encoded {
				t.Fatalf("\nwant %s\n got %s", fixture.Encoded, encoded)
			}
			framed := make([]byte, 4+len(encoded))
			copy(framed[4:], encoded)
			framed[0] = byte(len(encoded) >> 24)
			framed[1] = byte(len(encoded) >> 16)
			framed[2] = byte(len(encoded) >> 8)
			framed[3] = byte(len(encoded))
			if hex.EncodeToString(framed[:4]) != fixture.TCPPrefixHex {
				t.Fatal("prefix drift")
			}
		})
	}
	for _, fixture := range document.Invalid {
		t.Run(fixture.Name, func(t *testing.T) {
			var err error
			if fixture.Kind == "call" {
				_, err = DecodeCall([]byte(fixture.Encoded))
			} else {
				_, err = DecodeReceipt([]byte(fixture.Encoded))
			}
			if err == nil {
				t.Fatalf("accepted invalid fixture %s", fixture.Encoded)
			}
		})
	}
}

func TestReceiptStateMachineAndCorrelation(t *testing.T) {
	invalid := [][]byte{
		[]byte(`{"v":1,"op":"receipt","id":"c","key":"get_item","ok":false}`),
		[]byte(`{"v":1,"op":"receipt","id":"c","key":"get_item","ok":true,"error":{"code":"bad"}}`),
		[]byte(`{"v":1,"op":"receipt","id":"c","key":"get_item","ok":false,"status":200,"error":{"code":"bad"}}`),
		[]byte(`{"v":1,"op":"receipt","id":"c","key":"get_item","ok":false,"status":500,"body":null,"error":{"code":"bad"}}`),
	}
	for _, payload := range invalid {
		if _, err := DecodeReceipt(payload); err == nil {
			t.Fatalf("accepted %s", payload)
		}
	}
	call := NewCall("c1", "get_item")
	result := SuccessReceipt("c2", "get_item", 200, OptionalJSON{})
	if err := AssertReceiptForCall(call, result); err == nil || !strings.Contains(err.Error(), "id") {
		t.Fatalf("expected id mismatch: %v", err)
	}
}

func TestNDJSONAndLengthPrefixAreBounded(t *testing.T) {
	call := NewCall("c1", "healthz")
	line, err := ToNDJSON(call)
	if err != nil {
		t.Fatal(err)
	}
	if !strings.HasSuffix(line, "\n") {
		t.Fatal("missing terminator")
	}
	if _, err := CallFromNDJSON([]byte(line + "{}\n")); err == nil {
		t.Fatal("accepted multiple lines")
	}
	framed, err := EncodeLengthPrefixed(call)
	if err != nil {
		t.Fatal(err)
	}
	partial := append(append([]byte{}, framed...), 0, 0, 0)
	frames, rest, err := SplitLengthPrefixed(partial)
	if err != nil {
		t.Fatal(err)
	}
	if len(frames) != 1 || len(rest) != 3 {
		t.Fatalf("frames=%d rest=%d", len(frames), len(rest))
	}
	if _, _, err := SplitLengthPrefixed([]byte{0xff, 0xff, 0xff, 0xff}); err == nil {
		t.Fatal("accepted huge prefix")
	}
}

func TestCorrelatorIsConcurrentAndUnique(t *testing.T) {
	const count = 128
	correlator, err := NewCorrelator("rpc-")
	if err != nil {
		t.Fatal(err)
	}
	out := make(chan string, count)
	var group sync.WaitGroup
	for range count {
		group.Add(1)
		go func() {
			defer group.Done()
			value, err := correlator.Take()
			if err != nil {
				t.Errorf("Take: %v", err)
				return
			}
			out <- value
		}()
	}
	group.Wait()
	close(out)
	seen := map[string]struct{}{}
	for value := range out {
		if _, found := seen[value]; found {
			t.Fatalf("duplicate %s", value)
		}
		seen[value] = struct{}{}
	}
	if len(seen) != count {
		t.Fatalf("got %d", len(seen))
	}
	if _, err := NewCorrelator(strings.Repeat("x", 128)); err == nil {
		t.Fatal("accepted an unbounded correlation prefix")
	}
}

File: clients/typescript/src/rpc.test.js
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { test } from "node:test";
import { fileURLToPath } from "node:url";

import {
  Correlator,
  RpcV1Error,
  assertReceiptForCall,
  callFromNdjson,
  decodeCall,
  decodeReceipt,
  encodeCall,
  encodeLengthPrefixed,
  encodeReceipt,
  splitLengthPrefixed,
  toNdjson,
} from "./rpc.js";

const here = dirname(fileURLToPath(import.meta.url));
const fixtures = JSON.parse(
  readFileSync(join(here, "../../../examples/rpc-v1/conformance.json"), "utf8"),
);
const hex = (bytes) =>
  [...bytes].map((value) => value.toString(16).padStart(2, "0")).join("");

test("fixture profile is v1 and not RIDL v2", () => {
  assert.equal(fixtures.profile, "ores-rpc-v1-call-receipt");
  assert.equal(fixtures.schemaVersion, 1);
});

for (const fixture of fixtures.valid) {
  test(`${fixture.name} round-trips exact shared bytes`, () => {
    const envelope = fixture.kind === "call"
      ? decodeCall(fixture.encoded)
      : decodeReceipt(fixture.encoded);
    assert.equal(toNdjson(envelope).slice(0, -1), fixture.encoded);
    assert.equal(
      hex(encodeLengthPrefixed(envelope).subarray(0, 4)),
      fixture.tcp_prefix_hex,
    );
  });
}

for (const fixture of fixtures.invalid) {
  test(`${fixture.name} fails closed`, () => {
    assert.throws(
      () => fixture.kind === "call"
        ? decodeCall(fixture.encoded)
        : decodeReceipt(fixture.encoded),
      RpcV1Error,
    );
  });
}

test("constructors enforce the receipt state machine", () => {
  const call = encodeCall({ id: "c1", key: "healthz", transport: "tcp" });
  const success = encodeReceipt({
    id: "c1",
    key: "healthz",
    transport: "tcp",
    ok: true,
    status: 200,
    body: null,
  });
  assert.equal(success.body, null);
  assert.equal(assertReceiptForCall(call, success), success);
  assert.throws(
    () => encodeReceipt({ id: "c1", key: "healthz", ok: false, status: 500 }),
    /needs error/,
  );
  assert.throws(
    () => encodeReceipt({
      id: "c1",
      key: "healthz",
      ok: true,
      status: 200,
      error: { code: "bad" },
    }),
    /must not carry error/,
  );
});

test("correlation mismatches fail closed", () => {
  const call = encodeCall({ id: "c1", key: "healthz", transport: "tcp" });
  const receipt = encodeReceipt({
    id: "c2",
    key: "healthz",
    transport: "tcp",
    ok: true,
    status: 200,
  });
  assert.throws(() => assertReceiptForCall(call, receipt), /id does not match/);
});

test("NDJSON accepts one terminator and rejects multiple objects", () => {
  assert.equal(
    callFromNdjson('{"v":1,"op":"call","id":"c1","key":"healthz"}\r\n').id,
    "c1",
  );
  assert.throws(
    () => callFromNdjson(
      '{"v":1,"op":"call","id":"c1","key":"healthz"}\n{}\n',
    ),
    /exactly one JSON object/,
  );
});

test("length-prefix decoder rejects oversized declarations and keeps a tail", () => {
  const first = encodeLengthPrefixed(encodeCall({ id: "c1", key: "healthz" }));
  const second = encodeLengthPrefixed(encodeReceipt({
    id: "c1",
    key: "healthz",
    ok: true,
    status: 200,
  }));
  const buffer = new Uint8Array(first.length + 3);
  buffer.set(first);
  buffer.set(second.subarray(0, 3), first.length);
  const { frames, rest } = splitLengthPrefixed(buffer);
  assert.equal(frames.length, 1);
  assert.equal(rest.length, 3);
  assert.throws(
    () => splitLengthPrefixed(new Uint8Array([0xff, 0xff, 0xff, 0xff])),
    /over the .* limit/,
  );
});

test("invalid UTF-8, unknown members, symbols, and cycles fail closed", () => {
  assert.throws(() => decodeCall(new Uint8Array([0xff])), /UTF-8/);
  const symbol = Symbol("hidden");
  assert.throws(
    () => encodeCall({ id: "c1", key: "healthz", [symbol]: true }),
    /member names must be strings/,
  );
  const cyclic = {};
  cyclic.self = cyclic;
  assert.throws(
    () => encodeCall({ id: "c1", key: "healthz", body: cyclic }),
    /cyclic/,
  );
});

test("correlation identifiers are monotonic and bounded", () => {
  const correlator = new Correlator("request-");
  assert.deepEqual([correlator.take(), correlator.take()], ["request-1", "request-2"]);
  assert.throws(() => new Correlator("\ud800"), /Unicode scalar/);
});
